### PR TITLE
Add back in the LambdaSimplifier analyzer+fixer.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is now built on top of diagnostic analyzers and IOperation.  It will now work on VB as well.

### CODE REVIEW

**Comment by jmarolf** [comment]
:+1:

File: src/Features/CSharp/Portable/CodeRefactorings/MoveDeclarationNearReference/MoveDeclarationNearReferenceCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -174,7 +174,7 @@ private StatementSyntax CreateMergedDeclarationStatement(State state, StatementS
                 // Type inference.  Only merge if types match.
                 var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
                 var rightType = semanticModel.GetTypeInfo(expression.Right, cancellationToken);
-                return (localSymbol.Type == null && rightType.Type == null) || localSymbol.Type.Equals(rightType.Type);
+                return Equals(localSymbol.Type, rightType.Type);
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeRefactorings/MoveDeclarationNearReference/MoveDeclarationNearReferenceCodeRefactoringProvider.cs]: Wow, I kind of hate this.  Would seem much clearer to me to have `object.Equals` in this case.

File: src/EditorFeatures/CSharpTest/CodeActions/LambdaSimplifier/LambdaSimplifierTests.cs [hunk 2]
```diff
@@ -236,9 +242,9 @@ static void M()
 
             await TestAsync(code, expected, compareTokens: false);
         }
-
+        
         [WorkItem(545856, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/545856")]
-        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsLambdaSimplifier)]
+        [Fact(Skip = "NotYetImplemented"), Trait(Traits.Feature, Traits.Features.CodeActionsLambdaSimplifier)]
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/CodeActions/LambdaSimplifier/LambdaSimplifierTests.cs]: This needs an active bug, or should just be removed.

File: src/EditorFeatures/CSharpTest/CodeActions/LambdaSimplifier/LambdaSimplifierTests.cs [hunk 3]
```diff
@@ -1,51 +1,57 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/CodeActions/LambdaSimplifier/LambdaSimplifierTests.cs]: No tests for VB?  Bad @CyrusNajmabadi.  No :cookie: for you.

File: src/Features/Core/Portable/LambdaSimplifier/LambdaSimplifierDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            }
+
+            var firstStatement = body.Statements[0];
+            return lambdaExpression.Signature.ReturnsVoid
+                ? TryGetInvocationExpressionForSubLambda(firstStatement)
+                : TryGetInvocationExpressionForFuncLambda(firstStatement);
+        }
+
+        private static IInvocationExpression TryGetInvocationExpressionForSubLambda(
+            IOperation operation)
```
Comment (Pilchie) [inline src/Features/Core/Portable/LambdaSimplifier/LambdaSimplifierDiagnosticAnalyzer.cs]: I don't love this pair of methods, but maybe it's just that I haven't used `IOperation` enough yet.  Do we really need to Unwrap conversions everywhere?  Is that safe in these contexts, or is it changing meaning?

**Comment by Pilchie** [comment]
@CyrusNajmabadi Do you want to resolve conflicts and add some VB tests here, or is that blocked on IOperation work ATM? (I guess it would also need to be rebased to `future` for the IOperation dependency).

**Comment by CyrusNajmabadi** [comment]
This is blocked on IOperation work.  Specifically: https://github.com/dotnet/roslyn/issues/10488  and https://github.com/dotnet/roslyn/pull/10486

**Comment by jaredpar** [comment]
We apologize, but we are closing this PR due to code drift. We regret letting this PR go so long without attention, but at this point the code base has changed too much for us to revisit older PRs. Going forward, we will manage PRs better under an SLA currently under draft in issue #26266 – we encourage you to add comments to that draft as you see fit.   Although that SLA is still in draft form, we nevertheless want to move forward with the identification of older PRs to give contributors as much early notice as possible to review and update them.

If you are interested in pursuing this PR, please reset it against the head of master and we will reopen it. Thank you!

**Comment by CyrusNajmabadi** [comment]
@jaredpar can you reopen.  I do want this to go in.

**Comment by CyrusNajmabadi** [comment]
there was no work on this because one of the issues i filed on this has not been addressed.  It seems poor form to close a PR out because it is blocked on other bugs that still have not been responded to :)

**Comment by jaredpar** [comment]
The PR's been inactive for almost two years at this point. I understand it's blocked on other work. At the same time though we're drowning in PRs. It's making it difficult for us to keep track of PRs that do and don't need attention here.  Once the other work is unblocked we should be able to activate this PR and move forward.

**Comment by CyrusNajmabadi** [comment]
> The PR's been inactive for almost two years at this point

No it hasn't.  :)  I updated it back in february to be up to date.  Just because something was created years ago, and has been kept up to date doesn't mean it should be closed out.  :)

**Comment by jaredpar** [comment]
Ah okay. I missed that last block of commits when I was looking through the history here.

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by CyrusNajmabadi** [comment]
closing

**Pull request closed without merging**